crates/oxc_semantic/tests/conformance/test_symbol_declaration.rs

```diff
@@ -0,0 +1,108 @@
+/// 3. The binding pattern or identifier in that node should be populated (e.g. not [`None`]) and
+///    contain the symbol id.
+///
+/// [`SymbolTable`]: oxc_semantic::SymbolTable
+#[derive(Debug, Clone, Default)]
+pub struct SymbolDeclarationTest;
+
+/// The binding pattern or identifier contained in the declaration node is [`None`].
+///
+/// See: [`BindingIdentifier::symbol_id`]
+fn bound_to_statement_with_no_binding_identifier(
+    symbol_id: SymbolId,
+    span: Span,
+    statement_kind: &str,
+) -> TestResult {
+    OxcDiagnostic::error(format!(
+        "Symbol {symbol_id:?} got bound to a {statement_kind} with no BindingIdentifier"
+    ))
+    .with_label(span.label("Symbol was declared here"))
+    .into()
+}
+
+/// [`BindingIdentifier::symbol_id`] contained [`Some`] value, but it was not the [`SymbolId`] used
+/// to find it in the [`SymbolTable`].
+fn symbol_declaration_not_in_ast_node(
+    expected_id: SymbolId,
+    binding: &BindingIdentifier,
+) -> TestResult {
+    let bound_id = binding.symbol_id.get();
+    OxcDiagnostic::error(format!(
+        "Expected binding to be bound to {expected_id:?} but it was bound to {bound_id:?}"
+    ))
+    .with_label(binding.span())
+    .into()
+}
+
+/// Found a non-destructuring [`BindingPattern`] that did not contain a [`BindingIdentifier`].
+fn malformed_binding_pattern(expected_id: SymbolId, pattern: &BindingPattern) -> TestResult {
+    OxcDiagnostic::error(format!("BindingPattern for {expected_id:?} is not a destructuring pattern but get_binding_identifier() still returned None"))
+        .with_label(pattern.span().label("BindingPattern is here"))
+        .into()
+}
+
+impl ConformanceTest for SymbolDeclarationTest {
+    fn name(&self) -> &'static str {
+        "symbol-declaration"
+    }
+
+    fn run_on_symbol(
+        &self,
+        symbol_id: oxc_semantic::SymbolId,
+        semantic: &Semantic<'_>,
+    ) -> TestResult {
+        let declaration_id = semantic.symbols().get_declaration(symbol_id);
+        let declaration = semantic.nodes().get_node(declaration_id);
+        let span = semantic.symbols().get_span(symbol_id);
+
+        match declaration.kind() {
+            AstKind::VariableDeclarator(decl) => check_binding_pattern(symbol_id, &decl.id),
+            AstKind::CatchParameter(caught) => check_binding_pattern(symbol_id, &caught.pattern),
+            AstKind::Function(func) => match func.id.as_ref() {
+                Some(id) => check_binding(symbol_id, id),
+                None => bound_to_statement_with_no_binding_identifier(symbol_id, span, "Function"),
+            },
+            AstKind::Class(class) => match class.id.as_ref() {
+                Some(id) => check_binding(symbol_id, id),
+                None => bound_to_statement_with_no_binding_identifier(symbol_id, span, "Class"),
+            },
+            AstKind::FormalParameter(param) => check_binding_pattern(symbol_id, &param.pattern),
+            _ => TestResult::Pass,
```

Comment:
Shouldn't this be a test fail not pass?

Reply:
There are other declarations I didn't get around to. I'll add them.